permuta/meshpatt.py

```diff
@@ -867,238 +698,117 @@ def ascii_plot(self, cell_size=1):
-            number: <numbers.Integral>
-                An integer of which binary representation corresponds to a
-                legal shading.
-        Raises:
-            TypeError:
-                Bad argument type.
-            ValueError:
-                Bad argument, but correct type.
-        Examples:
-            >>> bin(22563)
-            '0b101100000100011'
-            >>> MeshPatt.unrank((0, 1, 2), 386)
-            MeshPatt(Perm((0, 1, 2)), [(0, 1), (1, 3), (2, 0)])
-        """
-        if not isinstance(number, numbers.Integral):
-            message = "'{}' object is not an integer".format(repr(number))
-            raise TypeError(message)
-        if not (0 <= number < 2 ** ((len(pattern) + 1) ** 2)):
-            message = "Element out of range: '{}'".format(number)
-            raise ValueError(message)
-        bound = len(pattern) + 1
-        shading = set(
-            (index // bound, index % bound)
-            for index, bit in enumerate(reversed(bin(number)[2:]))
-            if bit == "1"
-        )
-        return MeshPatt(pattern, shading)
-
-    @staticmethod
-    def random(length):
-        """Return a random mesh pattern of the specified length.
+        n, tab = len(self), " " * 4
+        lis = [
+            "\\begin{tikzpicture}",
+            f"[scale=.3,baseline=(current bounding box.center)]\n{tab}",
+            f"\\foreach \\x in {{1,...,{n}}} {{\n{tab*2}",
+            f"\\draw[ultra thin] (\\x,0)--(\\x,{n+1}); %vline\n{tab*2}",
+            f"\\draw[ultra thin] (0,\\x)--({n+1},\\x); %hline\n{tab}}}",
+            "".join(
+                (
+                    f"\n{tab}\\fill[pattern color = black!75, pattern="
+                    f"north east lines] {cell} rectangle +(1,1);"
+                )
+                for cell in sorted(self.shading)
+            ),
+            f"\n{tab}",
+            f"\n{tab}".join(
+                f"\\draw[fill=black] ({idx+1},{val+1}) circle (5pt);"
+                for idx, val in enumerate(self.pattern)
+            ),
+            "\n\\end{tikzpicture}",
+        ]
+        return "".join(lis)
 
-        Args:
-            length: <numbers.Integral>
-                The length of the random pattern.
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, self.__class__):
+            return self.shading == other.shading and self.pattern == other.pattern
+        return False
 
-        Examples:
-            >>> mp = set(MeshPatt.unrank(Perm((0, )), i) for i in range(0, 16))
-            >>> MeshPatt.random(1) in mp
-            True
-            >>> len(MeshPatt.random(4))
-            4
-        """
-        return MeshPatt.unrank(
-            Perm.random(length), random.randint(0, 2 ** ((length + 1) ** 2) - 1)
-        )
+    def __hash__(self) -> int:
+        return hash((self.pattern, self.shading))
 
-    #
-    # Dunder methods
-    #
+    def __iter__(self) -> Iterator[Union[Perm, FrozenSet]]:
```

Comment:
Agreed but this was inheriting from a `NamedTuple` before and some tests failed until I added `__iter__` so someone is using `MeshPatt` this way. It might just be in tests but I didn't want to change to much in other parts so I suggest we keep it for now.